### PR TITLE
added socket_path_template parameter

### PR DESCRIPTION
**Pull request by Vladikamira**
Hi, guys!

MoM's project is great! I want to use it with my virtualization system Opennebula! 
But unfortunately, it has own socket storing location based on `Libvirt Domain ID` and `Opennebula Vm ID`
The path looks like that:
```
[root@virthost.dev:~]# ls -l /var/lib/libvirt/qemu/domain-132-one-151/monitor.sock 
srwxrwxr-x 1 oneadmin oneadmin 0 Jan  8 11:59 /var/lib/libvirt/qemu/domain-132-one-151/monitor.sock
```
so, I suggest to use templating for path like that
```
/var/lib/libvirt/qemu/channel/target/domain-%(domid)s-%(name)s
```
please, review changes and correct me if I made mistake or you have another idea how to solve it :)

### CODE REVIEW

**Comment by rollandf** [comment]
Hi,

Thank for the PR.
The github is just a mirror.
Please use Gerrit:
https://gerrit.ovirt.org/#/q/project:mom

Here an explanation on how to use Gerrit:
https://www.ovirt.org/develop/dev-process/working-with-gerrit/

Regards

**Comment by sandrobonazzola** [comment]
Hi @Vladikamira, was this submitted to gerrit.ovirt.org? If not, do you need help pushing it there?

**Comment by sandrobonazzola** [comment]
submitted to gerrit https://gerrit.ovirt.org/q/topic:%2522oVirt/mom/pull/1%2522

**Comment by nyoxi** [comment]
This has still not been merged and we are moving from Gerrit to GitHub so I am reopening the PR.

File: doc/mom-balloon+ksm.conf [hunk 1]
```diff
@@ -101,6 +115,14 @@ collectors: GuestQemuProc, GuestMemory, GuestBalloon
 #   after appropriate socket_name_template substitution as explained below.
 socket_path: /var/lib/libvirt/qemu
 
+# The same as socket_path variable, but in template format.
```
Comment (nyoxi) [inline doc/mom-balloon+ksm.conf:118]: Space at the end of line here and on the line below

File: doc/mom-balloon+ksm.conf [hunk 2]
```diff
@@ -101,6 +115,14 @@ collectors: GuestQemuProc, GuestMemory, GuestBalloon
 #   after appropriate socket_name_template substitution as explained below.
 socket_path: /var/lib/libvirt/qemu
 
+# The same as socket_path variable, but in template format. 
+# Some systems, like Opennebula store sockets in defferent folders, bassed on 
+# VM id and domain id. Here you can template it (example for opennebula):
+#   <socket_path_template>/<socket_name_template>
```
Comment (nyoxi) [inline doc/mom-balloon+ksm.conf:121]: I sort of understand what you are trying to express here (hint what opennebula should use), but in the context of the comment block this seems confusing. Could this be maybe formulated more explicitly (e.g.: Opennebula should use configuration values ...) or removed entirely?

File: doc/mom-balloon+ksm.conf [hunk 3]
```diff
@@ -101,6 +115,14 @@ collectors: GuestQemuProc, GuestMemory, GuestBalloon
...
+# Default: /var/lib/libvirt/qemu
+# socket_name_template: /var/lib/libvirt/qemu/channel/target/domain-%(domid)s-%(name)s
```
Comment (nyoxi) [inline doc/mom-balloon+ksm.conf:123]: Should this be `socket_path_template`? Also please remove the leading space.

File: doc/mom-balloon+ksm.conf [hunk 4]
```diff
@@ -101,6 +115,14 @@ collectors: GuestQemuProc, GuestMemory, GuestBalloon
...
+# Note: you should comment socket_path variable first, or template will be ignored
```
Comment (nyoxi) [inline doc/mom-balloon+ksm.conf:124]: Please move this line up (possibly above the default value) so that the example is the last line in the block.

Alternatively, keep this line where it is but move the line with example above the default, replacing the "<socket_path_template>/<socket_name_template>"

File: doc/mom-balloon.conf [hunk 5]
```diff
@@ -101,6 +115,14 @@ collectors: GuestQemuProc, GuestMemory, GuestBalloon
 #   after appropriate socket_name_template substitution as explained below.
 socket_path: /var/lib/libvirt/qemu
 
+# The same as socket_path variable, but in template format.
```
Comment (nyoxi) [inline doc/mom-balloon.conf:118]: Same comments as in the other file apply to this block.

File: mom/Collectors/GuestQemuAgent.py [hunk 6]
```diff
@@ -43,7 +46,21 @@ def __init__(self, properties):
...
+                 socket_path = '/var/lib/libvirt/qemu'
+
+        try:
+            socket_path
+        except NameError:
+            try:
+                 socket_path = socket_path_template % {'name' : self.name,'domid' : properties['domid']}
+            except KeyError, e:
+                 socket_path = socket_path_template
+                 self.logger.warn("Error substituting socket path " \
```
Comment (nyoxi) [inline mom/Collectors/GuestQemuAgent.py:61]: Backlash at the end of this line is not necessary.

File: mom/Collectors/GuestQemuAgent.py [hunk 7]
```diff
@@ -43,7 +46,21 @@ def __init__(self, properties):
...
+                                 "template. Invalid key: %s" % e)
+                 self.logger.warn("Using socket path %s." % socket_path)
```
Comment (nyoxi) [inline mom/Collectors/GuestQemuAgent.py:63]: Maybe this should be turned into info level and printed every time and not just in case of error.

File: mom/Collectors/GuestQemuAgent.py [hunk 8]
```diff
@@ -168,17 +186,26 @@ def collect(self):
             self.swap_in_prev = self.swap_in_cur
         if self.swap_out_prev is None:
             self.swap_out_prev = self.swap_out_cur
-        swap_in = self.swap_in_cur - self.swap_in_prev
-        swap_out = self.swap_out_cur - self.swap_out_prev
 
-        data = { 'mem_available': avail, 'mem_unused': unused, \
-                 'mem_free': free, 'swap_in': swap_in, 'swap_out': swap_out, \
+        if self.swap_in_cur is None:
```
Comment (nyoxi) [inline mom/Collectors/GuestQemuAgent.py:190]: Where did these changes come from? How are they related?

File: mom/Graphite.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# Memory Overcommitment Manager
+# Copyright (C) 2010 Adam Litke, IBM Corporation
```
Comment (nyoxi) [inline mom/Graphite.py:2]: This copyright is wrong -- wrong year, wrong copyright holder. Please replace it, idealy with something generic like:

```
# Copyright oVirt Authors
# SPDX-License-Identifier: GPL-2.0-only
```

File: mom/Graphite.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
```
Comment (nyoxi) [inline mom/Graphite.py:15]: If you use the SPDX licence identifier please remove the above license lines.

File: doc/mom-balloon.conf [hunk 11]
```diff
@@ -39,6 +39,20 @@ libvirt-hypervisor-uri:
 # created and updated with all data generated by the configured Collectors.
 plot-dir:
 
+# Set Graphite host to collect all metrics there
+# Metrics are same as collected for 'plot-dir'
+# Metrics will be collected in path: <FQDN>.mom.metric1
```
Comment (nyoxi) [inline doc/mom-balloon.conf:44]: Space at the end of the line.

File: doc/mom-balloon.conf [hunk 12]
```diff
@@ -39,6 +39,20 @@ libvirt-hypervisor-uri:
 # created and updated with all data generated by the configured Collectors.
 plot-dir:
 
+# Set Graphite host to collect all metrics there
+# Metrics are same as collected for 'plot-dir'
+# Metrics will be collected in path: <FQDN>.mom.metric1 
+# disabled by default
+graphite-host:
```
Comment (nyoxi) [inline doc/mom-balloon.conf:46]: Space at the end of the line.

File: mom/Collectors/GuestQemuAgent.py [hunk 13]
```diff
@@ -133,7 +150,7 @@ def getfile(self, path, maxSize=1048576):
         while True:
             ret = self.agent_cmd('file_read', fh, 1024)
             data += ret['buf']
-            if len(ret) < 1024:
+            if len(ret['buf']) < 1024:
```
Comment (nyoxi) [inline mom/Collectors/GuestQemuAgent.py:153]: How is this change related?

File: mom/Collectors/GuestQemuAgent.py [hunk 14]
```diff
@@ -155,6 +172,7 @@ def collect(self):
         swap_total = parse_int("^SwapTotal: (.*) kB", meminfo)
         swap_free = parse_int("^SwapFree: (.*) kB", meminfo)
         free = unused + buffers + cached
+        used = avail - free
```
Comment (nyoxi) [inline mom/Collectors/GuestQemuAgent.py:175]: How is this change related?

File: mom/Collectors/GuestQemuAgent.py [hunk 15]
```diff
@@ -168,17 +186,26 @@ def collect(self):
...
                  'major_fault': majflt, 'minor_fault': minflt, \
                  'swap_total': swap_total, \
                  'swap_usage': swap_total - swap_free }
         return data
 
     def getFields(self=None):
-        return set(['mem_available', 'mem_unused', 'mem_free',
-                    'major_fault', 'minor_fault', 'swap_in', 'swap_out',
-                    'swap_total', 'swap_usage'])
+        return set(['mem_available', 'mem_unused', 'mem_used', 'mem_free',
```
Comment (nyoxi) [inline mom/Collectors/GuestQemuAgent.py:209]: Again, how is this related?

File: mom/Graphite.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+class Graphite:
+    def __init__(self, graphite_host, graphite_port, graphite_protocol, name):
+    
+      self.graphite_host = graphite_host
+      self.graphite_port = graphite_port
+      self.graphite_protocol = graphite_protocol
+      self.vmname = name
+
+      if graphite_protocol == "TCP":
+         self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (nyoxi) [inline mom/Graphite.py:31]: We need to set a reasonably small timeout (`self.sock.settimeout()`) for the operations on the socket to avoid blocking the monitor. The common interval for collectors is 5 seconds so let's go with this value. But we are talking about overall timeout for the operation, so see also the other comment below.

File: mom/Graphite.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+import socket
+import logging
+
+class Graphite:
+    def __init__(self, graphite_host, graphite_port, graphite_protocol, name):
+    
+      self.graphite_host = graphite_host
+      self.graphite_port = graphite_port
+      self.graphite_protocol = graphite_protocol
+      self.vmname = name
```
Comment (nyoxi) [inline mom/Graphite.py:28]: `vmname` is misleading. It could also be `HostMonitor` for host monitor. Maybe better to use something like `monitor_name`.

File: mom/Graphite.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+      if graphite_protocol == "TCP":
+         self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+         try:
+            self.sock.connect((graphite_host, graphite_port))
+         except Exception as e:
```
Comment (nyoxi) [inline mom/Graphite.py:34]: You assume `self.socket` is `None` if there is no connection, but the object is never invalidated. I suppose it should be done here.

File: mom/Graphite.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+            logger = logging.getLogger('mom.Graphite')
+            logger.warn("something's wrong with %s:%d. Exception is %s" % (graphite_host, graphite_port, e))
```
Comment (nyoxi) [inline mom/Graphite.py:36]: Any reason why not use `logger.exception()` here?

File: mom/Graphite.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+    def sendToGraphite(self, data):
+        if self.sock is None:
+            return
+
+        for key,value in data.items():
+           if self.graphite_protocol == "TCP":
+              timestamp = int(time.time())
+              gr_path = "direct." + socket.getfqdn() + ".mom." + self.vmname + "." + key
+              self.sock.send("%s %d %d\n" % (gr_path, int(value), timestamp))
```
Comment (nyoxi) [inline mom/Graphite.py:54]: We need to limit the time for all the `send()` calls in the block. One way would be to divide the overall timeout by number of values in the data collection. Probably a better way would be (assuming the protocol allows it) to gather all the messages in one string and perform just one `send()` operation.

File: mom/Graphite.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+           elif self.graphite_protocol == "UDP":
+              timestamp = int(time.time())
+              gr_path = "direct." + socket.getfqdn() + ".mom." + self.vmname + "." + key
+              self.sock.sendto( gr_path + " " + str(value) + " " + str(timestamp) + "\n" , (self.graphite_host, self.graphite_port) )
```
Comment (nyoxi) [inline mom/Graphite.py:58]: It seems to me the only difference is in the method used (`send` vs. `sendto`) so this `if` block could be simplified, especially in the light of the previous comment.

File: mom/Graphite.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+            self.sock.close()
+
+    def sendToGraphite(self, data):
+        if self.sock is None:
+            return
+
+        for key,value in data.items():
+           if self.graphite_protocol == "TCP":
+              timestamp = int(time.time())
+              gr_path = "direct." + socket.getfqdn() + ".mom." + self.vmname + "." + key
```
Comment (nyoxi) [inline mom/Graphite.py:53]: I don't think `socket.getfqdn` does any caching. Are you sure it cannot do reverse-DNS lookup? One way or the other, calling it on every iteration of the for-loop may not be the best idea.

File: mom/Makefile.am [hunk 23]
```diff
@@ -31,6 +31,7 @@ mom_PYTHON = \
 	MOMFuncs.py \
 	Monitor.py \
 	Plotter.py \
+        Graphite.py \
```
Comment (nyoxi) [inline mom/Makefile.am:34]: Wrong indentation, need to use TAB character.

**Review by nyoxi — COMMENTED** [review]
Hello, apologies it took us so long to notice this and thank you very much for the contribution.

There are some code issues that need to be resolved before the patches can be accepted. Also there are some changes that seem unrelated -- possibly caused by wrong rebase or wrong base branch. Finally please squash the changes into a single commit since we do not do merge commits. Alternatively consider seprating the changes logically into commits which would help the review process. To give you some ideas it could be:

1) socket path template
2) graphite support

or maybe:

1) socket path tempalte
2) basic Graphite support with TCP protocol
3) UDP protocol

Thank you again for your contribution.

**Comment by sandrobonazzola** [comment]
@Vladikamira can you please rebase and resolve conflicts?
